### ISSUE TITLE
Frozen Pydantic model is reported as not hashable.

### ISSUE THREAD

**Issue by mgsimpson-i2** [microsoft/pyright#6481]
**Describe the bug**
When a Pydantic model is frozen, Pyright does not consider it to be hashable.

**Code or Screenshots**

With this code in "foo.py":
```python
from pydantic import (
    BaseModel,
    ConfigDict,
)

class FrozenModel(BaseModel):
    pass

    model_config = ConfigDict(
        frozen=True,
    )

m = FrozenModel()
d = {
    m: "foo",
}

print(f"value for key m in dict d is \"{d[m]}\"")
```

Python runs clean:
```
% python foo.py
value for key m in dict d is "foo"
```

Pyright reports an error:
```
% pyright foo.py
/vagrant/packages/i2ns-api/foo.py
  /vagrant/packages/i2ns-api/foo.py:15:5 - error: Dictionary key must be hashable
    Type "FrozenModel" is not hashable (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
```

Using:
- Python 3.12.0
- Pydantic 2.4.2
- Pyright 1.1.336

**Miscellaneous Notes**
- This may be a recurrence of a previously-fixed issue https://github.com/microsoft/pyright/issues/2784.
- I believe that Pydantic 2.x changed the way it handles the "frozen" designation internally, which may be why it recurred (original issue was against Pydantic 1.x, I believe).

**Comment by erictraut** [comment]
Pyright has no specific knowledge of pydantic or any other third-party library. It bases its analysis on the type information provided by the library author. Pydantic uses the `dataclass_transform` mechanism defined in [PEP 681](https://peps.python.org/pep-0681/) to indicate that `BaseModel` should act like a stdlib `dataclass`. 

The `model_config` mechanism used by pydantic is non-standard and specific to pydantic, so it's not handled by `dataclass_transform`. 

You may want to reach out to the maintainers of pydantic to see if they have any suggested workarounds.

I'm going to close this issue because pyright is working as designed here.

**Comment by erictraut** [comment]
Here's a potential workaround:

```python
if TYPE_CHECKING:

    @dataclass_transform(frozen_default=True)
    class FrozenModel:
        ...

else:

    class FrozenModel(BaseModel):
        pass

        model_config = ConfigDict(
            frozen=True,
        )
```
